File: infrastructures/ddd/python/src/infrastructure/repositories/infra/i_query_repository.py
"""
Base interface for query repositories (read operations in CQRS).
Query repositories work with ReadEntities and return them directly to service.
No BMO mapping in read path for performance.
"""

from abc import ABC, abstractmethod
from typing import Generic, List, Optional, TypeVar


TReadEntity = TypeVar('TReadEntity')
TId = TypeVar('TId')


class IQueryRepository(ABC, Generic[TReadEntity, TId]):
    """
    Base interface for query repositories (read operations).
    
    Query repositories:
    - Work with ReadEntities
    - Return ReadEntity directly to service (no BMO mapping)
    - Handle query operations (find, list, search)
    - Optimized for read performance
    """
    
    @abstractmethod
    async def find_by_id_async(self, p_id: TId) -> Optional[TReadEntity]:
        """
        Finds an entity by ID.
        Returns ReadEntity directly (no BMO in read path).
        
        Args:
            p_id: The identifier to find
            
        Returns:
            The read entity or None if not found
        """
        pass
    
    @abstractmethod
    async def list_by_filter_async(
        self,
        p_skip: int = 0,
        p_take: int = 10
    ) -> List[TReadEntity]:
        """
        Lists entities with pagination.
        Returns ReadEntities directly.
        
        Args:
            p_skip: Number of records to skip
            p_take: Number of records to take
            
        Returns:
            List of read entities
        """
        pass

File: infrastructures/ddd/typescript/src/infrastructure/repositories/infra/IQueryRepository.ts
/**
 * Base interface for query repositories (read operations in CQRS).
 * Query repositories:
 * - Return ReadEntities directly to services
 * - No BMO mapping in read path for performance
 * - Optimized for query operations
 * - Work with denormalized read models
 * 
 * @template TReadEntity The read entity type
 * @template TId The identifier type
 */
export interface IQueryRepository<TReadEntity, TId> {
  /**
   * Finds an entity by its identifier.
   * Returns ReadEntity directly (no BMO mapping).
   * @param p_id The identifier to find
   * @returns The read entity or null if not found
   */
  FindByIdAsync(p_id: TId): Promise<TReadEntity | null>;

  /**
   * Lists entities with pagination.
   * Returns ReadEntities directly (no BMO mapping).
   * @param p_skip Number of records to skip
   * @param p_take Number of records to take
   * @returns Array of read entities
   */
  ListAsync(p_skip: number, p_take: number): Promise<TReadEntity[]>;

  /**
   * Counts total number of entities.
   * @returns The total count
   */
  CountAsync(): Promise<number>;
}
